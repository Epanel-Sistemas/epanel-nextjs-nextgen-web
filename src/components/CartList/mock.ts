export default [
  {
    id: '1',
    title: 'The Legend of Adventure',
    price: '$10.50'
  },
  {
    id: '2',
    title: 'Galactic Odyssey',
    price: '$10.50'
  }
]
