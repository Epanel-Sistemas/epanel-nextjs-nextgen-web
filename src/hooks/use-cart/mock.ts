export const gamesMock = [
  {
    id: '1',
    name: 'The Legend of Adventure',
    price: 10.5
  },
  {
    id: '2',
    name: 'Galactic Odyssey',
    price: 10.5
  }
]

export const cartItems = [
  {
    id: '1',
    price: '$10.50',
    title: 'The Legend of Adventure'
  },
  {
    id: '2',
    price: '$10.50',
    title: 'Galactic Odyssey'
  }
]
