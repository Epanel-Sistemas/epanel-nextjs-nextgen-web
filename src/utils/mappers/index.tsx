import formatPrice from '~/utils/format-price'

import { GameProps } from '~/types/api'

export const cartMapper = (games: GameProps[] | undefined) => {
  return games
    ? games.map((game) => ({
        id: game.id,
        title: game.name,
        price: formatPrice(game.price)
      }))
    : []
}
