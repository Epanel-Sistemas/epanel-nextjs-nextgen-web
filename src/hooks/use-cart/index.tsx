import { useContext, createContext, useState, useEffect } from 'react'

import formatPrice from '~/utils/format-price'
import { getStorageItem, setStorageItem } from '~/utils/storage'
import { cartMapper } from '~/utils/mappers'
import { gamesMock } from './mock'

const CART_KEY = 'cartItems'

export type CartItem = {
  id: string
  title: string
  price: string
}

export type CartContextData = {
  items: CartItem[]
  quantity: number
  total: string
  isInCart: (id: string) => boolean
  addToCart: (id: string) => void
  removeFromCart: (id: string) => void
  clearCart: () => void
}

export const CartContextDefaultValues = {
  items: [],
  quantity: 0,
  total: '$0.00',
  isInCart: () => false,
  addToCart: () => null,
  removeFromCart: () => null,
  clearCart: () => null
}

export const CartContext = createContext<CartContextData>(
  CartContextDefaultValues
)

export type CartProviderProps = {
  children: React.ReactNode
}

const CartProvider = ({ children }: CartProviderProps) => {
  const [cartItems, setCartItems] = useState<string[]>([])

  useEffect(() => {
    const data = getStorageItem(CART_KEY)

    if (data) {
      setCartItems(data)
    }
  }, [])

  const total = gamesMock.reduce((acc, game) => {
    return acc + game.price
  }, 0)

  const isInCart = (id: string) => (id ? cartItems.includes(id) : false)

  const saveCart = (cartItems: string[]) => {
    setCartItems(cartItems)
    setStorageItem(CART_KEY, cartItems)
  }

  const addToCart = (id: string) => {
    saveCart([...cartItems, id])
  }

  const removeFromCart = (id: string) => {
    const newCartItems = cartItems.filter((itemId: string) => itemId !== id)
    saveCart(newCartItems)
  }

  const clearCart = () => {
    saveCart([])
  }

  return (
    <CartContext.Provider
      value={{
        items: cartMapper(gamesMock),
        quantity: cartItems.length,
        total: formatPrice(total),
        isInCart,
        addToCart,
        removeFromCart,
        clearCart
      }}
    >
      {children}
    </CartContext.Provider>
  )
}

const useCart = () => useContext(CartContext)

export { CartProvider, useCart }
