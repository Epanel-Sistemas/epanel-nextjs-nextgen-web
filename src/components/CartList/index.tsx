import Link from 'next/link'
import CartItem from '~/components/CartItem'

import { Button } from '~/components/ui/button'
import Empty from '~/components/Empty'

import { useCart } from '~/hooks/use-cart'

export type CartListProps = {
  hasButton?: boolean
}

const CartList = ({ hasButton = false }: CartListProps) => {
  const { items, total } = useCart()

  return (
    <div className="flex flex-col gap-2" data-cy="cart-list">
      {items.length ? (
        <>
          <div className="flex flex-col gap-2">
            {items.map((item) => (
              <CartItem key={item.title} {...item} />
            ))}
          </div>

          <div>
            {!hasButton && <span>Total:</span>}
            <span>{total}</span>

            {hasButton && (
              <Link href="/cart" passHref>
                <Button>Buy it now</Button>
              </Link>
            )}
          </div>
        </>
      ) : (
        <Empty
          title="Your cart is empty"
          description="Go back to the store and explore great games and offers."
        />
      )}
    </div>
  )
}

export default CartList
