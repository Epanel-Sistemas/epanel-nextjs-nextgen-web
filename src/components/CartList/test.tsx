import { CartContextDefaultValues } from '~/hooks/use-cart'
import { render, screen } from '~/utils/test-utils'

import CartList from '.'
import items from './mock'

describe('<CartList />', () => {
  it('should render the cart list', () => {
    const cartProviderProps = {
      ...CartContextDefaultValues,
      items,
      total: 'R$ 330,00'
    }

    const { container } = render(<CartList />, { cartProviderProps })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the button', () => {
    const cartProviderProps = {
      ...CartContextDefaultValues,
      items
    }

    render(<CartList hasButton />, { cartProviderProps })

    expect(screen.getByText(/buy it now/i)).toBeDefined()
  })

  it('should render empty if there are no games', () => {
    render(<CartList />)

    expect(screen.getByText(/your cart is empty/i)).toBeDefined()
  })
})
