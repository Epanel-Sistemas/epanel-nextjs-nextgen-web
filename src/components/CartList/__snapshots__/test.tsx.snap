// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartList /> should render the cart list 1`] = `
<div
  class="flex flex-col gap-2"
  data-cy="cart-list"
>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="flex flex-col gap-2"
      data-cy="game-item"
    >
      <div
        class="flex flex-col gap-2"
      >
        <span>
          The Legend of Adventure
        </span>
        <div>
          <span>
            $10.50
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex flex-col gap-2"
      data-cy="game-item"
    >
      <div
        class="flex flex-col gap-2"
      >
        <span>
          Galactic Odyssey
        </span>
        <div>
          <span>
            $10.50
          </span>
        </div>
      </div>
    </div>
  </div>
  <div>
    <span>
      Total:
    </span>
    <span>
      R$ 330,00
    </span>
  </div>
</div>
`;
